File: data-structure/linked/doubly_linked.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        last = self.head
        while last.next:
            last = last.next
        last.next = new_node
        new_node.prev = last

    def print_list(self):
        current = self.head
        while current:
            print(f"{current.data} <-> ", end="")
            current = current.next
        print("NULL")

# 使用示例
dll = DoublyLinkedList()
dll.append(1)
dll.append(2)
dll.append(3)
dll.print_list()  # 输出：1 <-> 2 <-> 3 <-> NULL

"""
jarry@MacBook-Pro linked % python doubly_linked.py 
1 <-> 2 <-> 3 <-> NULL
"""

File: 01-data-structures/linked/doubly_linked.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLinkedList {
    constructor() {
        this.head = null;
    }

    append(data) {
        const newNode = new Node(data);
        if (this.head === null) {
            this.head = newNode;
            return;
        }
        let last = this.head;
        while (last.next !== null) {
            last = last.next;
        }
        last.next = newNode;
        newNode.prev = last;
    }

    printList() {
        let current = this.head;
        let result = '';
        while (current !== null) {
            result += current.data + ' <-> ';
            current = current.next;
        }
        console.log(result + 'NULL');
    }
}

const list = new DoublyLinkedList();
list.append(1);
list.append(2);
list.append(3);
list.printList(); // 输出：1 <-> 2 <-> 3 <-> NULL

/*
jarry@MacBook-Pro linked % node doubly_linked.js
1 <-> 2 <-> 3 <-> NULL
*/